### DISCUSSION TITLE
Relax Csla0011 for Commands

### DISCUSSION THREAD

**Discussion by ajj7060** [MarimerLLC/csla#1771] [Ideas]
I see this was discussed here: https://github.com/MarimerLLC/cslaforum/issues/817

I like the idea of this analyzer generally, and I know we can fix it by changing the ctor to a standard method and adding attributes, but we have hundreds of command objects and we have enough warnings to deal with, so to update from 4 to 5 I ended up just turning it off completely.

All of our command usages follow this:

```
public static MyCommand DoIt(string whatever) => DataPortal.Execute(new MyCommand(whatever));

public MyCommand() { }
private MyCommand(string whatever) => Whatever = whatever;
```

Its not clear to me what the benefits of changing the ctor to a method and using `DataPortal` to create the command instance are, and it adds an opportunity to do the wrong thing if the `[RunLocal]` gets forgotten.  I don't think we've ever encountered a need to actually hit the server to initialize a command just before we go to execute it.  

If the command doesn't require any parameters to execute, you're then forced to create an empty create method as well, again with a chance to forget `[RunLocal]`

CriteriaBase already seems exempt and it has managed properties as well, so it doesn't seem like allowing ctors for commands should be any kind of technical issue.  

Thoughts?

**Comment by rockfordlhotka** [comment]
When @JasonBock wrote the analyzer we discussed whether it should apply to all the base types that support stereotypes, and I thought it should.

I know what you are talking about @ajj7060, as there is certainly work associated with adding dummy `Create` methods for each command type, and that's not productive.

Certainly one answer would be to relax the analyzer. Another would be for the data portal (at least with a command type) to not require an explicit `RunLocal` `Create` operation to be declared. Maybe that should be the behavior for all types?

**Reply by ajj7060** [reply]
If the default for Create is RunLocal, then there probably needs to be a RunRemote attribute to force a remote call when it actually is desired, especially if that ends up being the default behavior for Create on all types.  That might be surprising for someone that switching from always local data portal to a remote one, but I think BusinessBase's parameter-less DataPortal_Create had that applied by default also so maybe that's expected.  I do like the consistency of Csla 5 always honoring the local/remote config by default.

I do think part of my issue is that the DataPortal opens a seam where the code could run remotely or locally, depending, but in this particular case initializing the command is more similar to initializing a criteria object, where that seam doesn't seem appropriate (to me, at least).  

Could this analyzer be split?  0011 works against bb/blb/dlb/rob/rolb and the new one looks at commandbase?  Of course I'm not familiar with the implications of that, so maybe that's not a good idea.

**Reply by rockfordlhotka** [reply]
The whole criteria concept is essentially gone now, so using that as a basis for anything isn't something I want to do. Now that the data portal takes multiple parameters, it is hard to see why anyone would use that older concept.

My thought with the "default create" wasn't to change the behavior for a `Create` method _you implement_, but rather to default to `RunLocal` if there is no `Create` method implementation at all.

**Reply by ajj7060** [reply]
Well I wasn't trying to say that criteriabase does it so command should to; its more the work it would require to satisfy this anaylzer, which for now has forced me to shut it off.  

The comment about the criteriabase being obsolete b/c the dataportal can take multiple arguments is interesting though; there's only `DataPortal.Execute<T>(T obj)` still.

Now if you could do:
`DataPortal.Execute<T>(param1, param2)`

and have it end up in
`[Execute] private void Execute(string param1, int param2)` 

That would eliminate the need for a `ctor` or a `[Create]` method.   Then the only properties on the command, if any, would be outputs of the command.

**Reply by rockfordlhotka** [reply]
Oh, I like the idea of passing parameters to the command like that - interesting idea!

**Reply by rockfordlhotka** [reply]
@ajj7060 would you please add that as a feature suggestion in the Issues tab?

**Reply by rockfordlhotka** [reply]
If @JasonBock figured out how to serialize a tuple (I think you were playing with that at one point?) we could have a new "command base" that accepts parameters and returns a tuple as a result. That might be pretty cool!

```
var resultAsTuple = DataPortal.Execute<T>(param1, param2);
```

**Reply by ajj7060** [reply]
> @ajj7060 would you please add that as a feature suggestion in the Issues tab?

Done.  You can find it here:  https://github.com/MarimerLLC/csla/issues/1775

Let me know if it captures the idea or if there's anything to tweak.
